### PR TITLE
I2C and External Interrupt Controller Drivers for Atmel's SAMD21 MCU

### PR DESCRIPTION
**Pull request by madlain**
Hi,

I've implemented a basic I2C driver and an external interrupt controller driver for Atmel's SAMD21 MCU. I've performed functional testing for both by exercising the sample application for the FXAS21002 gyroscope running on an Arduino MKRZero board connecting to Adafruits NXP Precision 9DoF breakout board.

Madani.

### CODE REVIEW

**Review by dbkinder — APPROVED** [review]
+1 for doc, thanks!

**Comment by dbkinder** [comment]
... but you do have lots of issues with coding style noted in the shippable details above :(

**Comment by madlain** [comment]
I'm not sure why the shippable run successive to the merge conflict resolution is failing. Can you please help me out?

**Comment by madlain** [comment]
Is there anything needed from me to move forward with the merge?

**Comment by galak** [comment]
@anangl can you review the i2c driver.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Will you please add the new board to `samples/sensor/fxas21002/sample.yaml`? This will ensure that the i2c driver gets built in CI.

File: boards/arm/arduino_mkrzero/arduino_mkrzero.dts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	chosen {
+		zephyr,console = &sercom5;
+		zephyr,sram = &sram0;
+		zephyr,flash = &flash0;
+	};
+};
+
+&sercom0 {
+	status = "ok";
+	compatible = "atmel,sam0-i2c";
```
[the remark this answers is not in the document]
Comment (madlain) [reply boards/arm/arduino_mkrzero/arduino_mkrzero.dts]: I've adhered to the exact same approach taken for the arduino_zero also based on Atmel SAMD21: the SoC device tree file only declares SERCOM devices while the board device tree file specifies how a given SERCOM is configured (USART, SPI or I2C) for that board. Is there anything wrong with this approach?

File: drivers/gpio/gpio_sam0.c [hunk 2]
```diff
@@ -9,10 +10,167 @@
...
+};
+
+#ifdef CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS
+
+DEVICE_DECLARE(gpio_sam0_0);
+
+#define GPIO_SAM0_PORTA_EXTINT(n) \
+	{ CONFIG_GPIO_SAM0_PORTA_EXTINT_##n##_PIN, n },
+
+static struct extint porta_extints[] = {
```
[the remark this answers is not in the document]
Comment (madlain) [reply drivers/gpio/gpio_sam0.c]: The porta_extints array allows using designated initializers for both the
mapping table and its length in the const config struct. Moving the definition
of the table to the const config struct prevents using a designated initializer
for the length and defers its computation to run time, using pointer arithmetic
in the device init function.

File: drivers/gpio/gpio_sam0.c [hunk 3]
```diff
@@ -113,13 +366,187 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
...
+{
+	int i;
+	u32_t pin_mask;
+	u32_t clear_mask;
+	struct device *dev;
+	const struct gpio_sam0_config *config;
+	struct gpio_sam0_data *data;
+	u32_t enabled_int;
+
+#ifdef CONFIG_GPIO_SAM0_PORTA_BASE_ADDRESS
```
[the remark this answers is not in the document]
Comment (madlain) [reply drivers/gpio/gpio_sam0.c]: Not sure how this macro could help here. This code is in the Interrupt Service
Routine for the External Interrupt Controller where the only way to get a
reference to the structure for the GPIO ports is to use their respective device
names with DEVICE_GET.

File: boards/arm/arduino_mkrzero/arduino_mkrzero_defconfig [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+# CONFIG_GPIO_SAM0_PORTB_EXTINT_2_PIN=2
+
+# FXAS21002C gyroscope I2C address on the Adafruit Precision NXP 9-DOF breakout
+# board
+# CONFIG_FXAS21002_I2C_NAME="SERCOM0"
+# CONFIG_FXAS21002_I2C_ADDRESS=0x21
+# CONFIG_FXAS21002_RANGE=3
+
+# CONFIG_FXAS21002_GPIO_NAME="PORTB"
+# CONFIG_FXAS21002_GPIO_PIN=2
```
[the remark this answers is not in the document]
Comment (madlain) [reply boards/arm/arduino_mkrzero/arduino_mkrzero_defconfig]: The board doesn't carry an gyro so I've removed  (commented out) all configs including the range. The reason why these were present is that, **at the time**, it was possible to build the sample app for the sensor with an **external gyro board** hooked up to the MKRZero. This set-up provided a convenient way to exercise the newly developed I2C and EIC drivers all together with the already existing drivers for the SoC (UART, GPIO, PINMUX).

File: boards/arm/arduino_mkrzero/doc/arduino_mkrzero.rst [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
++-----------+------------+--------------------------------------+
+| SYSTICK   | on-chip    | systick                              |
++-----------+------------+--------------------------------------+
+| WDT       | on-chip    | Watchdog                             |
++-----------+------------+--------------------------------------+
+| GPIO      | on-chip    | I/O ports                            |
++-----------+------------+--------------------------------------+
+| USART     | on-chip    | Serial ports                         |
++-----------+------------+--------------------------------------+
+| SPI       | on-chip    | Serial Peripheral Interface ports    |
```
[the remark this answers is not in the document]
Comment (madlain) [reply boards/arm/arduino_mkrzero/doc/arduino_mkrzero.rst]: I've enabled and tested the USART, GPIO, and PINMUX drivers.

**Comment by madlain** [comment]
Hi,

Thanks for the review and the valuable comments. I’ve addressed and resolved all but three of them for which I’m expecting your feedback. Also as clarification, and to answer your top-level comment, the board doesn't carry the FXAS21002 gyro. At the time I implemented the I2C and EIC drivers, it was possible to build the sample app for the sensor with an external gyro board hooked up to the MKRZero. This set-up provided a convenient way to exercise the newly developed I2C and EIC drivers all together with the already existing drivers for the SoC (UART, GPIO, PINMUX).

**Comment by jfischer-phytec-iot** [comment]
SAM0 I2C driver was introduced by #14128, @madlain please rebase or close

**Comment by galak** [comment]
Closing PR as its been marked stale for several months.  Feel free to re-open and update if desired.

**Pull request closed without merging**